api.go

```diff
@@ -25,84 +25,142 @@ import (
 	"mime/multipart"
 	"net"
 	"net/http"
+	"net/url"
 	"os"
+	"path"
 	"strconv"
 	"time"
 )
 
+const (
+	httpTimeout = 60 * time.Second
+	timeout     = 30 * time.Second
+	apiBasePath = "/api/v1"
+	regURL      = "/devices"
+	authURL     = "/authenticate_device"
+)
+
+type CacophonyDevice struct {
+	group    string
+	name     string
+	password string
+}
+
 type CacophonyAPI struct {
+	device         *CacophonyDevice
+	httpClient     *http.Client
 	serverURL      string
-	group          string
-	name           string
-	typeName       string
-	regURL         string
-	authURL        string
-	password       string
 	token          string
 	justRegistered bool
-	client         *http.Client
+}
+
+// joinURL creates an absolute url with supplied baseURL, and all paths
+func joinURL(baseURL string, paths ...string) string {
+
+	u, err := url.Parse(baseURL)
+	if err != nil {
+		return ""
+	}
+	url := path.Join(paths...)
+	u.Path = path.Join(u.Path, url)
+	return u.String()
+}
+
+func (api *CacophonyAPI) getAPIURL() string {
+	return joinURL(api.serverURL, apiBasePath)
+}
+
+func (api *CacophonyAPI) getAuthURL() string {
+	return joinURL(api.serverURL, authURL)
+}
+func (api *CacophonyAPI) getRegURL() string {
+	return joinURL(api.serverURL, apiBasePath, regURL)
 }
 
 func (api *CacophonyAPI) Password() string {
-	return api.password
+	return api.device.password
 }
 
 func (api *CacophonyAPI) JustRegistered() bool {
 	return api.justRegistered
 }
 
-const httpTimeout = 60 * time.Second
-const timeout = 30 * time.Second
-const deviceName = "devicename"
-const basePath = "/api/v1"
+// NewAPIFromConfig prases the supplied configFile and creates a new CacophonyAPI with the configFile information
+// and saves the generated password to privConfigFileName(configFile)
+func NewAPIFromConfig(configFile string) (*CacophonyAPI, error) {
+	conf, err := ParseConfigFile(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("configuration error: %v", err)
+	}
+	privConfigFilename := privConfigFilename(configFile)
+	password, err := ReadPassword(privConfigFilename)
+	if err != nil {
+		return nil, err
+	}
+
+	api, err := NewAPI(conf.ServerURL, conf.Group, conf.DeviceName, password)
+	if err != nil {
+		return nil, err
+	}
+
+	if api.JustRegistered() {
+		err := WritePassword(privConfigFilename, api.Password())
+		if err != nil {
+			return nil, err
+		}
+	}
+	return api, nil
```

Comment:
thinking about it some more... you want to take a lock before ReadPassword and hold it until after WritePassword might have run because if there is no password and NewAPIFromConfig is called concurrently then you can end up with 2 clients trying to register with different passwords at the same time.